Visual Script Graph Unification

swarnimarun opened the pull request:
This is part of my GSOC project.
This is still heavily WIP and needs probably a lot of refactoring work.

Here's a quick showcase,

![vs-showcase](https://user-images.githubusercontent.com/19930870/59276709-73e91800-8c7c-11e9-8ca6-63e68153b450.gif)

cc: @fire @KidRigger 


## Bugs
- [x] Get Copy & Paste of Nodes working.
- [x] Ensure minimum collision of bounding boxes of functions on import of scripts from old versions.
- [x] Undo not working properly with the connection rescans.
- [x] Quick Create Functions shortcut (Ctrl + G)
- [x] Proper Mapping of Returns from functions.
- [x] Update function calls on function argument change or function rename.
- [x] Add simple warnings for all unintended use cases.
- [x] Test Undo edge cases().

### Old bugs
- [x] NodePath edit with empty scene crashes the editor.
- [x] Make nodes Duplication work for connections as well.

### Additional Changes (UI)
- https://github.com/godotengine/godot/pull/29996 Better Type Detection
- https://github.com/godotengine/godot/pull/29995 Improve search for Available Nodes
- Add Port Swapping

fire commented:
Not placing it directly on top of another node is important, especially since we know the exact positions. I'll review more.

fire commented on modules/visual_script/visual_script.cpp:
```diff
@@ -2042,11 +2056,14 @@ void VisualScriptInstance::create(const Ref<VisualScript> &p_script, Object *p_o
 
 	for (const Map<StringName, VisualScript::Function>::Element *E = script->functions.front(); E; E = E->next()) {
 
+		if(E->key() == "fn_312592")
```
Huh? What is this for?

swarnimarun replied:
This is the name I used for the default function(basically the graph itself) for now... 

I planned to change to something built-in inside of the Visual Script and use the get_graph or something from script to get the values but it works fine enough at the moment, and keeping the whole thing using the same function system is much easier to allow for maintaining compatibility.

swarnimarun replied:
Technically it's a hack but using a weird name for default_func is actually a pretty ok solution. I need to move it to the Visual Script function itself. :P

So don't have to just copy paste strings.

fire replied:
Are you able to create a separate list for the default_func? The scary part of this is if the other side can guess this function name, they're able to replace the entire graph.

swarnimarun replied:
Are we trying to prevent hacking now?

But let them guess it, (it not human normal people don't use it) and even if they do it doesn't ever get executed. I will even remove it's last few fragments of it being added to the Godot Runtime.

swarnimarun replied:
Are we trying to prevent hacking now?

But let them guess it, (it not human normal people don't use it) and even if they do it doesn't ever get executed. I will even remove it's last few fragments of it being added to the Godot Runtime.

swarnimarun replied:
Are we trying to prevent hacking now?

But let them guess it, (it not human normal people don't use it) and even if they do it doesn't ever get executed. I will even remove it's last few fragments of it being added to the Godot Runtime.

swarnimarun commented:
@fire Yep, I haven't added any UI changes at the moment just finishing the functionality at the moment.

swarnimarun commented:
It's in a proper working state now, just needs more testing to improve some more UI aspects if needed.
And also needs a clean up before it becomes ready for review.

![vs-show](https://user-images.githubusercontent.com/19930870/59534439-f1768780-8f0b-11e9-9030-1adebc6541e6.gif)

cc: @fire @KidRigger if any of you want certain specific changes do let me know.

swarnimarun commented:
A couple more gifs to show additions,
![vs-group_datas](https://user-images.githubusercontent.com/19930870/59723009-d7f27a00-9242-11e9-92b8-36409c35e0b9.gif)
![vs-grouping](https://user-images.githubusercontent.com/19930870/59723010-d88b1080-9242-11e9-80e0-8e9fe90c3b9c.gif)

fire commented:
There are some clang format problems.

```*** The following differences were found between the code to commit and the clang-format rules:
--- a/modules/visual_script/visual_script_editor.h	2019-06-21 20:29:57.059454622 +0000
+++ b/modules/visual_script/visual_script_editor.h	2019-06-21 20:30:04.981732182 +0000
@@ -193,7 +193,7 @@
 	void _end_node_move();
 	void _move_node(StringName func, int p_id, const Vector2 &p_to);
 
-	void _get_ends(int node, const List<VisualScript::SequenceConnection> &seqs, const Set<int> &selected, Set<int>& end_nodes);
+	void _get_ends(int node, const List<VisualScript::SequenceConnection> &seqs, const Set<int> &selected, Set<int> &end_nodes);
 
 	void _node_moved(Vector2 p_from, Vector2 p_to, int p_id);
 	void _remove_node(int p_id);
*** Aborting, please fix your commit(s) with 'git commit --amend' or 'git rebase -i <hash>'```

swarnimarun commented:
Yep somehow I managed to get my clang format settings messed up.. Fixed next PR should fix it all.

swarnimarun commented:
Other relevant PRs: https://github.com/godotengine/godot/pull/29995 https://github.com/godotengine/godot/pull/29996

fire commented:
Travis is still failing.

swarnimarun commented:
Shadowing is a problem... :sigh:

swarnimarun commented:
Demo of the UI changes:
![vs-swap](https://user-images.githubusercontent.com/19930870/59982236-a60b5a00-962c-11e9-9091-571e156fed6b.gif)

zohozer commented:
I’m curious if it is possible to implement a very useful feature from Bolt Unity into Godot Visual node system? I would really love to see the Flow Graph arrows like in Bolt, so to have a real visual feedback of what is happening at any time between the nodes.

swarnimarun commented:
@zohozer Not sure if it's possible at the moment as it would need Live Sync and also Live Editing support but will give it a look if it's easy enough no problems in adding more expressiveness :)

fire commented:
There are some formatting errors.

swarnimarun commented:
New Demo for Compose Array Node:
![vs-list-nodes](https://user-images.githubusercontent.com/19930870/60760792-110b5680-a059-11e9-9949-1a01e03d757f.gif)

Starting to move the Side Bar to the Top.(Can also be seen in the Gif)

swarnimarun commented:
Right Click for Nodes,
![vs-right-click](https://user-images.githubusercontent.com/19930870/61003775-44642300-a382-11e9-806e-ee69320a8c52.gif)

zombietype reviewed (commented):
Looks clean. Will build later in the day and check.

fire commented:
Did you try?

```diff
diff --git a/modules/visual_script/visual_script_editor.cpp b/modules/visual_script/visual_script_editor.cpp
index 37adb7575..4694078a4 100644
--- a/modules/visual_script/visual_script_editor.cpp
+++ b/modules/visual_script/visual_script_editor.cpp
@@ -1144,7 +1144,7 @@ void VisualScriptEditor::_create_function() {
 void VisualScriptEditor::_add_node_dialog() {
 	// available_nodes_dialog->popup();
 	// available_nodes_dialog->set_position(graph->get_global_position() + Vector2(55, 80));
-	_generic_search("", graph->get_global_position() + Vector2(55, 80));
+	_generic_search(script->get_instance_base_type(), graph->get_global_position() + Vector2(55, 80));
 }
 
 void VisualScriptEditor::_node_item_selected() {
@@ -1597,7 +1597,7 @@ void VisualScriptEditor::_graph_gui_input(const Ref<InputEvent> &p_event) {
 		Point2 gpos = Input::get_singleton()->get_mouse_position();
 		// available_nodes_dialog->popup();
 		// available_nodes_dialog->set_position(gpos);
-		_generic_search("", gpos);
+		_generic_search(script->get_instance_base_type(), gpos);
 	}
 }
 
```

![Screen Shot 2019-07-18 at 11 46 33 PM](https://user-images.githubusercontent.com/32321/61515012-519e9480-a9b6-11e9-90af-63bd357dc517.png)

This enables the show current node feature.

fire commented:
Regarding the conversation in Discord I agree we should improve the "generic search" dialog.

> Tree refers to Godot Tree GUI where each Node is in it's own drop down.

As long as the search still works, it should be desired.

fire commented:
This is a picture of property search, but generic search should be able to get the script variables too. The "ball_dir" variable.

![Screen Shot 2019-07-19 at 12 16 16 AM](https://user-images.githubusercontent.com/32321/61516753-7b59ba80-a9ba-11e9-9746-6e7a1c325d8c.png)

fire commented:
We still can't construct objects in VS. https://github.com/godotengine/godot/issues/9067

fire commented:
I think we should prepare this pull request to be merged and make a separate pull request for vs group node.

fire commented:
Summary of meeting:

* We should try to get this PR https://github.com/godotengine/godot/pull/29681 merged. (Generic search, functionality in blog post and others.)
* Focus on fixing issues that are bugs and not feature requests on Github https://github.com/godotengine/godot/issues?q=is%3Aopen+is%3Aissue+label%3Atopic%3Avisualscript+label%3Abug
* Work on Group Nodes

fire commented on modules/visual_script/visual_script_editor.cpp:
```diff
@@ -997,6 +1117,44 @@ void VisualScriptEditor::_member_edited() {
+void VisualScriptEditor::_add_node_dialog() {
+	// available_nodes_dialog->popup();
+	// available_nodes_dialog->set_position(graph->get_global_position() + Vector2(55, 80));
+	_generic_search("", graph->get_global_position() + Vector2(55, 80));
```
```_generic_search(script->get_instance_base_type(), graph->get_global_position() + Vector2(55, 80));```

fire commented on modules/visual_script/visual_script_editor.cpp:
```diff
@@ -1319,22 +1568,37 @@ void VisualScriptEditor::_on_nodes_duplicate() {
+		Point2 gpos = Input::get_singleton()->get_mouse_position();
+		// available_nodes_dialog->popup();
+		// available_nodes_dialog->set_position(gpos);
+		_generic_search("", gpos);
```
```_generic_search(script->get_instance_base_type(), gpos);```

swarnimarun commented:
This PR is ready for review. 

I just need to remove a few more of the TODOs I have left in it.

Things I am not sure about,
- What should be the correct name for the base class for In-Graph Editable Nodes Class...(It's called VisualScriptLists atm)?
- If a dialog would be better for creating functions? (IMO without the dialog it feels better)

cc: @fire @KidRigger @akien-mga @Chaosus

fire commented:
Can you show a video of both with a dialog and without a dialog for creating functions?

What should be the correct name for the base class for In-Graph Editable Nodes Class...(It's called VisualScriptLists atm)? 

Can you describe more?

Chaosus commented:
![image](https://user-images.githubusercontent.com/3036176/61587417-cf1efc00-ab92-11e9-80d5-2c0818cfa063.png)

The dialog default size seems to be very small compared to the amount content it provides - also the size is resetted every time the user call it

swarnimarun commented:
I was confused about that, will make it longer, probably also give it some more width and I thought resizing was better... 
I guess it feels troublesome because it's too small...

swarnimarun commented on modules/visual_script/visual_script_nodes.h:
```diff
@@ -103,6 +103,103 @@ class VisualScriptFunction : public VisualScriptNode {
 	VisualScriptFunction();
 };
 
+class VisualScriptLists : public VisualScriptNode {
```
I was talking about this @fire I am thinking about improving the name...

fire replied:
Did you think of any ideas?

swarnimarun replied:
I think Lists is fine for now and considering it's a VirtualClass I don't think it will a breaking change even if we try to fix it later.

Chaosus commented:
IMO when user clicks on the function in the list - the graph should set the focus on that function

swarnimarun commented:
@Chaosus you can do that with Ctrl + Click I didn't make it the default as functions need to be editable from the view...

Chaosus commented:
Yeah, of course - i mean you can add a button to do it (its not obvious to use CTRL)

swarnimarun commented:
Makes sense..

swarnimarun commented:
@fire The function dialog would basically allow for a way to setup function on creation.

It would look similar to 
![image](https://user-images.githubusercontent.com/19930870/61596309-3b474180-ac1f-11e9-95bd-b0b03697f9c8.png)

PS: I stole the image from Chaosus' PR

fire commented:
The sidebar for adding arguments is quite difficult to use, so this is an improvement.

fire commented:
Some warnings:

```modules/visual_script/visual_script_editor.cpp:2347:14: error: declaration of 'gn' shadows a previous local [-Werror=shadow=local]
 2347 |   GraphNode *gn = Object::cast_to<GraphNode>(graph->get_child(i));
      |              ^~
modules/visual_script/visual_script_editor.cpp:2343:13: note: shadowed declaration is here
 2343 |  GraphNode *gn = Object::cast_to<GraphNode>(n);
      |             ^~```

fire commented:
![Screen Shot 2019-07-29 at 4 52 43 AM](https://user-images.githubusercontent.com/32321/62046431-c0e96500-b1bc-11e9-87dd-8c88c3982197.png)

Adding arguments and then deleting the arguments leaves empty space.

fire commented:
![Screen Shot 2019-07-29 at 4 55 03 AM](https://user-images.githubusercontent.com/32321/62046560-16be0d00-b1bd-11e9-8ee6-b65f1e987e15.png)


Is there a reason why there's ".." after add nodes?

fire commented:
Also the "Change Base Type:" is selectable, it shouldn't be.

Chaosus commented:
> Is there a reason why there's ".." after add nodes?

I guess for compatibility with visual shaders :) But to be honest there should be 3 dots after Add node (and better push this button at first place on the top bar like visual shader)

> Adding arguments and then deleting the arguments leaves empty space.

Yeah, to fix it simply add `control->set_size(Size2(-1, -1)) ` on deletion

swarnimarun commented:
I missed some dots and just noticed the resizing problem.. :P

fire commented:
No string input bug 

![image](https://user-images.githubusercontent.com/32321/62837461-8227c600-bc24-11e9-9a27-b8c53fe462f9.png)

Typing a character will make the search appear.

swarnimarun commented:
Screenshots of latest changes,
![image](https://user-images.githubusercontent.com/19930870/62996331-f317e980-be81-11e9-95b3-7f6833039132.png)
![image](https://user-images.githubusercontent.com/19930870/62996365-20649780-be82-11e9-8def-6510cca5b730.png)

fire commented:
How do you use the Fix FunctionCall Node bugfix?

swarnimarun commented:
@fire It should just work. Somehow magically, I swear I don't know why the FunctionCall Node was broken when you selected Instance mode.
I believe it's due to some commit in the upstream but I haven't checked the upstream on its own to know for sure.

swarnimarun commented:
Improved the Members Selection pane to be more intuitive.

![image](https://user-images.githubusercontent.com/19930870/64056314-f0223600-cbaf-11e9-86ee-53598596e89f.png)
- Double Click to center, right click or button to edit.

fire commented:
@swarnimarun Branch conflict

akien-mga commented on modules/visual_script/doc_classes/VisualScriptLists.xml:
```diff
@@ -0,0 +1,95 @@
+		<method name="add_input_data_port">
+			<return type="void">
+			</return>
+			<argument index="0" name="arg0" type="int" enum="Variant.Type">
```
You forgot to name arguments in the bindings.

swarnimarun replied:
Thanks for the catch. :)

akien-mga commented on modules/visual_script/visual_script_editor.h:
```diff
@@ -165,23 +170,27 @@ class VisualScriptEditor : public ScriptEditorBase {
+
+	Vector2 mouse_up_position;
+
+	void _port_action_menu(int p_option, const StringName &func);
```
Parameters should be prefixed with `p_`, so this should be `p_func`. You should review most changed or added prototypes, quite a few don't follow this convention.

swarnimarun replied:
Will do.

swarnimarun commented:
Sorry guys for the delay, I have been a little too busy with my mid-sem exams. I might be a little late on fixing up everything, should end in a few more days.
Sorry. :disappointed:

swarnimarun committed:
Visualscript editor graph unification & refactoring

Removes the need to have separate graphs per function for the VisualScript Nodes, and refactoring UI and other improvements such as fuzzy search, right click search boxes and in-graph editable nodes

triptych commented:
Hope this gets merged soon. Visual scripting needs these changes.

akien-mga commented:
Needs a rebase (probably due to the merge of the tool mode), otherwise I guess it should be good to merge :)

akien-mga commented:
Thanks a lot!

Estianne commented:
Please don't abandon this visual scripting part of godot

triptych commented:
I think this is a sign they are continuing to evolve it. I takes people passionate about the feature to help with filing bugs, and working on features for it to develop.
